libtiff/libtiff_ctypes.py

```diff
@@ -1953,16 +1972,16 @@ def suppress_warnings():
 def suppress_errors():
     libtiff.TIFFSetErrorHandler(_null_error_handler)
 
-
 def _test_custom_tags():
     def _tag_write():
         a = TIFF.open("/tmp/libtiff_test_custom_tags.tif", "w")
 
-        a.SetField("ARTIST", "MY NAME")
+        a.SetField("ARTIST", b"MY NAME")
```

Comment:
Does this mean that `.SetField` can only take bytes now? Otherwise it seems all the tests using plain strings are gone.

Reply:
Yes, that's correct, there is no way right now to store str directly in a TIFF tag. I think it's fine, Python strings are unicode, and TIFF doesn't have any official way to store unicode, only ASCII. So the only option to store Python strings is to encode them with a specific encoding. In practice, in our project, we always encode into utf-8, and I assume that nowadays a lot of software are able to handle it, but that's not standard. 

Maybe I could change that code to `"MY NAME".encode("ascii")`, to make it a little clearer. What do you think?

Reply:
Could the encoding done inside the SetField method? So that users can provide both `bytes` and `str` values. For the latter, we could also introduce `encoding="ascii"` kw argument to `SetField` if reasonable. What do you think?

Reply:
If we change that, I would suggest to do it in a separate pull request 😉

I think it would be relatively easy to encode str automatically to ASCII if the field is ASCII. It would accept either bytes or str. If it contains non ASCII characters, it would raise an exception, saying you have to pick the encoding yourself (or always encode to utf8?). It sounds like a nice helping behaviour.
However it would be hard to do in the other way around, in GetField, as we don't know what encoding is the data. So it could introduce a confusing API where you call SetField with str and GetField returns bytes.

Reply:
Notice that changing tests implies a possible BC breakage, in general.

I agree that for consistency the input and output types of `SetField` and `GetField` ought to match. IIUC, this PR makes `SetField` and `GetField` consistent wrt value types but it also removes a "feature" of `SetField` accepting `str` values.

I suggest that we keep the feature for now with a possible warning message that it is going to be removed in the future (in case we really want it) or with the warning message that if there are any issues with encoding/decoding str/bytes values, the recommended approach is to use `bytes` value in `SetField` (as `GetField` always returns a `bytes` value).

Reply:
Oh... there is a misunderstanding: this PR doesn't change the behaviour of SetField/GetField with respect to `str`. Currently, they don't accept `str`. As a matter of fact, all the test cases in libtiff_ctypes involving TIFF tags fail. (I think the breakage dates from the transition to Python 3).

This PR does 2 things: add support for extending with variable length tags *and* fix the test cases.
I'm sorry for the confusion. I started with adding the support for the variable length tags, and went on extending the test cases with such tags... and then realised that the all the test cases were broken in the first place, so I fixed them. I could separate these 2 changes in 2 commits or 2 PRs if you prefer.

So the good news is that no feature is removed: SetField and GetField have never accepted unicode strings as argument. It's just that in Python 2, bytes were called `str`.

Reply:
BTW @pearu , I've been thinking of moving all these test cases to the dedicated file `libtiff/tests/test_libtiff_ctypes.py`, which for now only contains a single, very lonely, test case. This way, they would be run in CI too. Do you agree that's the way to go?

Reply:
@pieleric thanks for the explanation, indeed, its not a BC breakage.
Re moving the tests: SGTM!

Reply:
Thanks for clarifying the situation @pieleric. I was definitely confused. Sounds good. Thanks.